src/Units/UnitModel.java

```diff
@@ -41,7 +41,8 @@
   public int idleFuelBurn;
   public int movePower;
   public int visionRange;
-  public boolean visionIgnoresCover = false;
+  public int visionRangePiercing = 1;
```

Comment:
All right, so I really don't like this. It absolutely feels like a kludge: fundamentally, units don't have two vision rages, and piercing vision isn't required to see adjacent units (whether cloaked or covered). In fact, on `master`, we don't have any unit or Commander who supports piercing vision at all.

I would go so far as to say we could (and probably should) implement the basics of visibility without needing this flag. Just delete it and get things working that way. If we overload the meaning or use of piercing vision it'll hurt us long-term.

Reply:
> fundamentally, units don't have two vision rages, and piercing vision isn't required to see adjacent units (whether cloaked or covered)

That's the thing; *all* units have piercing vision at 1 range or else you wouldn't be able to see into cover. The only difference is that "actual" piercing vision doesn't reveal stealthy units, but I plan to change that.
It's also worth noting that *properties* shouldn't get piercing vision (and they did until this change) (I might be willing to make an exception for units *on* them, but whatever)

> I would go so far as to say we could (and probably should) implement the basics of visibility without needing this flag. Just delete it and get things working that way.

I would do the exact same thing as I did before, but with a hard-coded "1" in the vision logic instead. The problem is that there's two different sight modes, *not* that both sight modes are configurable.

I *could* remove the second path by just making a new object with two collections, or having an in/out collection param, but that will still two passes, just through different collections.
I imagine it would also work to return, say, a `HashMap<XYCoord, bool>`, with the `keyset` being visible tiles and the `bool` determining whether cloaked units should be shown.

In essence, returning what *tiles* are revealed is no longer sufficient information to determine full visibility rules; some other flag is needed.

> If we overload the meaning or use of piercing vision it'll hurt us long-term.

What... what is this supposed to mean?

Reply:
> all units have piercing vision at 1 range or else you wouldn't be able to see into cover

I'd be more inclined to say that piercing vision isn't required to detect units at 1 range. Maybe it's a potato potato thing.

> > If we overload the meaning or use of piercing vision it'll hurt us long-term.
> What... what is this supposed to mean?

Piercing vision allows seeing enemies in cover when we wouldn't normally be able to. Units are normally able to see adjacent enemies, even when they are in cover. Using the piercing attribute to handle a case where piercing vision is not necessary is overloading the use of that attribute. Overloading things like that can lead to unintended dependencies or side effects. It may be clever, it may even work, but it normally means there is a better way to decompose the problem.
On the other hand, if we frame it as "all units have piercing vision at 1 range" then we could argue that it isn't being overloaded to handle this case.
On the other other hand, if all units have piercing vision at 1 range, it seems like we shouldn't need a separate UnitModel attribute for it.

Reply:
> I'd be more inclined to say that piercing vision isn't required to detect units at 1 range. Maybe it's a potato potato thing.

It's close, but not *quite* the same. Buildings have vision (and maybe piercing vision on top of themselves), but they should not be able to see into cover or detect at 1 range.
Whether you make a piercing vision stat or not, the fact would remain that *units* would have piercing vision at range 1.

> > > If we overload the meaning or use of piercing vision it'll hurt us long-term.
> Piercing vision allows seeing enemies in cover when we wouldn't normally be able to.

I would shorten this. Piercing vision allows seeing enemies in cover. I don't think it's helpful to make a special distinction between the vision you get of adjacent squares and "boosted" vision that lets you do the same feats of detection at higher ranges.

> Units are normally able to see adjacent enemies, even when they are in cover. Using the piercing attribute to handle a case where piercing vision is not necessary is overloading the use of that attribute.

Piercing vision, in my view, is the state of being able to detect hidden units. Units have it at range 1 by default, and a CO could conceivably grant it at greater ranges.
I don't see a good reason to special-case detection at 1 range. It's piercing vision, and works exactly the same as piercing vision.

> Overloading things like that can lead to unintended dependencies or side effects.

The only thing I can think of, here, is that hidden unit detection will *also* happen at whatever piercing vision range you have... but that's a good change, IMO.
Seeing as "piercing vision" and "can see into cover" are exactly equivalent in all respects... it's not overloading, it's looking at a duck and calling it a duck.

> On the other other hand, if all units have piercing vision at 1 range, it seems like we shouldn't need a separate UnitModel attribute for it.

Possibly not, but I don't really want to hardcode a *unit* attribute in some other piece of code.
Something I and others have contemplated in the past, for instance, is giving mobile SAMs piercing vision out to their normal vision range; this would give them a niche in being able to detect cloaked units without taking up attacking space (though that advantage is less with DoR fog rules).

Reply:
Hmmm. I think I will concede (for now, at least). At any rate, this is working now.

Reply:
I don't know how relevant this is to you, but I decided to doublecheck the behavior on DoR, and subs aren't revealed by moving past them (at least, outside of fog; flares can't "flare" outside fog either, it seems).

Reply:
So flares can't even use their ability if FoW is turned off? Strange. Not that you'd need to, but still.

Not too relevant; I don't have a strong stance on whether hidden/covered visibility conditions should match; it sorta makes sense that they would though.